### DISCUSSION TITLE
Properties ordering constraint

### DISCUSSION THREAD

**Discussion by sky-ruiroque** [apple/pkl#605] [Q&A]
Is there a way to guarantee that a given object has all properties alphabetically ordered?

For example:

```
features {
  aFeature = true
  bFeature = true
  cFeature = true
}
```

How can I do something that if I insert a new property that breaks the ordering:

```
features {
  aFeature = true
  dFeature = true
  bFeature = true
  cFeature = true
}
```

Then the evaluation will give an error, forcing me to sort the properties correctly:

```
features {
  aFeature = true
  bFeature = true
  cFeature = true
  dFeature = true
}
```

Is this possible?

**Comment by HT154** [comment]
Object properties are not inherently ordered in Pkl. Any reflection of definition order in evaluated output is an implementation detail. Off-hand, I can't come up with a viable way to assert on property order, even with using `pkl:reflect` on the table.

Can you say more about why this is desirable? Would you want this enforced on class/module declarations? At class/module amend sites? Both? If ordering is important, can you use a `Listing` instead?

**Reply by sky-ruiroque** [reply]
I'm trying to migrate a very large legacy configuration to Pkl. The problem with legacy configurations is not only the configuration itself but also the existing processes. The current 1500 lines JSON files are alphabetically ordered, and used by everyone like that to ease any visual inspection. We use Node.js scripts to enforce that sorting when creating or updating those files.

Since our output in Pkl is going to be those JSON files, I can still do the same on the JSON output, and sort the properties. But I'm trying to make the JSON and the Pkl files as close as possible to one another.

**Reply by HT154** [reply]
I played around with this a little and I think you can get what you want here, but it does rely on Pkl's internal implementation details around `Map` ordering. I can't promise this will always work, but it does as of 0.26.2.

```pkl
/// A class defined with properties not in alphabetical order
class Test {
    c: Int
    a: Int
    d: Int = a + b + c
    b: Int
}

/// A Mapping with entries not in alphabetical order
testMapping = new Mapping {
    ["b"] = 2
    ["a"] = 1
    ["c"] = 3
}

/// An instantiation of class [Test] with properties not in alphabetical order (and also different from the class definition)
testClass = new Test {
    b = 4
    c = 5
    a = 6
}

output {
    renderer = new JsonRenderer {
        converters {
            // !!! DANGER !!! - This relies on Pkl's implementation details around Map ordering
            [Typed] = (it) -> it.toMap().entries.sortBy((it) -> it.key).toMap((it) -> it.key, (it) -> it.value)
            [Mapping] = (it) -> it.toMap().entries.sortBy((it) -> it.key).toMap((it) -> it.key, (it) -> it.value)
            [Map] = (it) -> it.entries.sortBy((it) -> it.key).toMap((it) -> it.key, (it) -> it.value)
        }
    }
}
```

Evaluating this module results in:
```json
{
  "testClass": {
    "a": 6,
    "b": 4,
    "c": 5,
    "d": 15
  },
  "testMapping": {
    "a": 1,
    "b": 2,
    "c": 3
  }
}
```

Both the module top-level and the properties within the rendered class and `Mapping` have been sorted.
